### PR TITLE
#42 update eslint config

### PR DESCRIPTION
**Pull request by airondumael**
# Upgrade ESLint Config | rnfdigital/projects#42

Upgrade eslint and add new rules/options. 

Instead of just adding just the new rules I also added the rules that is not on our `eslint.js`. The rules that we don't want are turned "off".  I did this by comparing all the rules we have and the rules listed eslint [docs](http://eslint.org/docs/rules/template-tag-spacing). 

For more info on what's done and why see **Change Summary** below.

**NOTE**: all rules modified/added here are all suggestions. If you/team have different preference just comment below so we can all discuss. 

## TODO
- [X] add all available rules 
- [X] turn off the rules that we don't need
- [X] set options for newly added rules
- [X] update examples/tests
- [X] update changelog
- [X] bump version

### Added
- [class-methods-use-this](http://eslint.org/docs/rules/class-methods-use-this)
this is to notify us if we are creating methods that don't really need to be inside the class (can be a standalone function)
*needs list of allowed methods to not use this, if you know some comment below.*
- [no-global-assign](http://eslint.org/docs/rules/no-extra-label)
replacement for deprecated [no-native-reassign](http://eslint.org/docs/rules/no-native-reassign)
- [capitalized-comments](http://eslint.org/docs/rules/capitalized-comments)
helps make comments consistent though we rarely put comments on our code but still this will help in those rare cases. this will not affect eslint comments
- [func-call-spacing](http://eslint.org/docs/rules/func-call-spacing)
consistent function invocation. we always want the opening parenthesis to be next to the function being invoked not after a space nor after a newline.
- [func-name-matching](http://eslint.org/docs/rules/func-name-matching)
not that useful for us because we don't assign named functions to property or variable. I added this though just to be sure
- [lines-around-directive](http://eslint.org/docs/rules/lines-around-directive)
to enforce new line after directives
- [no-tabs](http://eslint.org/docs/rules/no-tabs)
spaces all the way
- [object-curly-newline](http://eslint.org/docs/rules/object-curly-newline)
helps on making our objects more consistent.
- [no-new-symbol](http://eslint.org/docs/rules/no-new-symbol)

### Removed
- [no-spaced-func](http://eslint.org/docs/rules/no-spaced-func)
deprecated: replaced by [func-call-spacing](http://eslint.org/docs/rules/func-call-spacing)
- [prefer-reflect](http://eslint.org/docs/rules/prefer-reflect)
deprecated

### Added but turned off
- [no-extra-label](http://eslint.org/docs/rules/no-extra-label)
not sure what labels are for
- [no-implicit-globals](http://eslint.org/docs/rules/no-implicit-globals)
this is for browser scripts not useful for codes that are module scoped
- [no-restricted-properties](http://eslint.org/docs/rules/no-restricted-properties)
can be useful but it's not a general rule
- [no-unmodified-loop-condition](http://eslint.org/docs/rules/no-unmodified-loop-condition)
we are not using loops and also this suggests modifying data inside loops which we are not allowed to do
- [no-unused-labels](http://eslint.org/docs/rules/no-unused-labels)
not sure what labels are for
- [no-restricted-globals](http://eslint.org/docs/rules/no-restricted-globals)
not sure what global variables to restrict if ever there is
- [max-lines](http://eslint.org/docs/rules/max-lines)
not sure what is the maximum acceptable lines per file
- [line-comment-position](http://eslint.org/docs/rules/line-comment-position)
not so useful as we rarely put comments on our codes.
- [multiline-ternary](http://eslint.org/docs/rules/multiline-ternary)
rule is not that good. if we use "always" short codes (`true ? 'shortStr' : 'anotherStr'`) will need to be separated to new line and if we use "never" ternary on JSX will not be readable.
- [no-plusplus](http://eslint.org/docs/rules/no-plusplus)
not needed
- [object-property-newline](http://eslint.org/docs/rules/object-property-newline)
we already have object curly newline
- [sort-keys](http://eslint.org/docs/rules/sort-keys)
there are some instances that we want to group properties together
- [no-confusing-arrow](http://eslint.org/docs/rules/no-confusing-arrow)
- [no-restricted-imports](http://eslint.org/docs/rules/no-restricted-imports)
- [yield-star-spacing](http://eslint.org/docs/rules/yield-star-spacing)
we don't use generators
- [one-var-declaration-per-line](http://eslint.org/docs/rules/one-var-declaration-per-line)
 [one-var](http://eslint.org/docs/rules/one-var) already handle this rule because we don't allow multiple declaration per var

### CODE REVIEW

**Comment by jetlawrence** [comment]
>capitalized-comments

regarding this rule, will the `eslint-disable-line`/`eslint-disable-next-line` be affected by this?

EDIT: my bad, just saw the examples in the docs, comments like these which are eslint specific are unaffected

**Comment by airondumael** [comment]
@jetlawrence , sorry forgot to mention it on the description. This rule will not affect eslint comments also `global` comments. You can see more info [here](http://eslint.org/docs/rules/capitalized-comments).

**Comment by DaleLJefferson** [comment]
Nice job @airondumael

**Review by peterallanmdavid — APPROVED** [review]
I like the object-curly-newline rule.

**Review by jetlawrence — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
   },
   "devDependencies": {
     "babel-eslint": "^7.1.1",
-    "eslint": "^3.14.1",
+    "eslint": "^3.16.1",
```
Comment (DaleLJefferson) [inline package.json:32]: Stupidly you need to keep the peerDependencies and devDependencies in sync, can you update the peerDependencies

**Review by DaleLJefferson — DISMISSED** [review]

**Review by jdmunro — APPROVED** [review]

**Pull request merged**